onchain-development intro-to-anchor updated (#349)
solana-foundation/developer-content: content/courses/onchain-development/intro-to-anchor.md

## Summary

- **Programs** on Solana have **instruction handlers**, which are functions that
  take arguments from incoming instructions. They are the entry point for any
  operation in a program.
- **Rust** is the most common language for building Solana programs. The
  **Anchor** framework takes care of common grunt work - like reading data from
  incoming instructions, and checking the right accounts are provided - so you
  can focus on building your Solana program.

## Lesson

Before we begin, make sure you have Anchor installed. You can follow this lesson

> on [local-setup](./local-setup.md).

Reviewer: This needs to be a full path, see 'links' in CONTRIBUTING.

> Solana's capacity to execute arbitrary code is a key part of its power. Solana
> programs, a.k.a "smart contracts", are the very foundation of the Solana

Reviewer: ```suggestion
Solana's capacity to execute arbitrary code is a key part of its power. Solana
programs (sometimes called "smart contracts"), are the very foundation of the Solana
```

Reviewer: We generally try to avoid smart contract and focus on programs, see CONTRIBUTING.

## Summary

- **Programs** on Solana have **instruction handlers** that execute instruction
  logic. They are the entry point for any operation in a program.
- **Rust** is the most common language for building Solana programs. The
  **Anchor** framework takes care of common grunt work - like reading data from
  incoming instructions, and checking the right accounts are provided - so you
  can focus on building your Solana program.

## Lesson

Solana's capacity to execute arbitrary code is a key part of its power. Solana
programs, akin to "smart contracts" in other blockchain systems, are the very
foundation of the Solana ecosystem. And as developers and creators continuously
conceive and deploy new programs, the collection of Solana programs continues to

> expand daily.

Reviewer: Honestly I think the old version of this sucks - it's full of vague terms like 'backbone' and 'powerful' and we can probably change it a lot more. 

> Solana programs, called "smart contracts", form much of the utility of Solana. Every popular Solana exchange, borrow-lend app, digital art auction house, perps platform, and prediction market is a program.

This lesson will give you a basic introduction to writing and deploying a Solana
program using the Rust programming language and the Anchor framework.

> This and the further lessons in this course will give a good base to start
> building Solana programs with Anchor, however if you want to get more into
> Anchor, we would recommend checking out the
> [The Anchor Book](https://book.anchor-lang.com/).

> 

Reviewer: Damn straight! We should have had a link to the book way earlier.

Anchor uses macros and traits to generate boilerplate Rust code for you. These
provide a clear structure to your program so you can more easily reason about
your code and focus more on the functionality of your program. The main

> high-level macros and attributes are:

Reviewer: Again this isn't you, I just think the old version of this is really vague and we can do better if we provide a more specific example: 

> Anchor uses macros and traits to generate boilerplate Rust code for you. For example, Solana programs using Anchor look like regular Rust modules - they use the `mod` keyword, and have functions inside. By applying the `program` macro to that module though, that regular Rust module becomes a Solana program - it will have an address, send incoming instructions to the right instruction handler function, and more. Anchor will generate the code to make the module a Solana program for us at compile time - we don't have to see that code or think about that code, we just apply the `program` macro and the work is done for us. 

Just a suggestion. Feel free to play with it!

> From here on out, you'll see a lot of Rust. We assume that you are familiar
> with Rust, if not, we recommend you to check out the
> [Rust Book](https://doc.rust-lang.org/book/).

- `declare_id!` - a macro for declaring the program’s onchain address
- `#[program]` - an attribute macro used to denote the module containing the
  program’s instruction logic. This is where all the instruction handlers are
  defined.
- `Accounts` - a trait applied to structs representing the list of accounts
  required for an instruction.
- `#[account]` - an attribute macro used to define custom account types for the
  program.

Let's talk about each of them before putting all the pieces together.

### Declare your program ID

> The `declare_id` macro sets the on-chain address of the Anchor program (i.e. the

Reviewer: `onchain` is one word. https://github.com/solana-foundation/developer-content/blob/main/CONTRIBUTING.md#prose

treated as a separate instruction.

Each instruction function requires a parameter of type `Context` and can

> optionally include additional function parameters representing instruction data
> needed by the handler to do the operation inside the instruction function.

Reviewer: ```suggestion
include more parameters as needed. 
```

Reviewer: Not you, but the old sentence was again necessary complicated.

}
```

- The `#[program]` attribute macro is used to denote the module containing the
  program’s instruction logic.
- `use super::*;` is used to bring all the items from the parent module into
  scope, which are needed to define the instruction logic.

> - Next, there is the instruction function definition. It's a simple public
>   function that writes some data (`instruction_data` in this case) to an
>   account.

Reviewer: ```suggestion
- Next, there is the instruction handler function. This function just writes some data (`instruction_data` in this case) to an account.
```

> }

Reviewer: Formatting above looks odd, see CONTRIUBUTING.md

Author: I think it only appears like this in the diff window. I checked locally and on my github branch, its appears to be properly formatted there.

Author: ![image](https://github.com/user-attachments/assets/8dd0bb49-aaa4-4508-8183-21e492a12f89)

Reviewer: I bet it's using a mix of tabs and spaces, and your local branch probably has some value set for tabs (2 spaces, 4 spaces, 8 spaces) that differs from what github uses. So yes it's a prob - please do fix it!

Author: Can you show/tell me whats exactly out of place here, I didn't touch this part and it seems like it is formatted like everything Rust part in the file. Also used `rustfmt` as instructed in CONTRIBUTING.md, nothing changes.

    /// Bump seeds found during constraint validation. This is provided as a
    /// convenience so that handlers don't have to recalculate bump seeds or
    /// pass them in as arguments.
    /// Type is the bumps struct generated by #[derive(Accounts)]
    pub bumps: T::Bumps,
}
```

`Context` is a generic type where `T` defines the list of accounts an
instruction requires. When you use `Context`, you specify the concrete type of
`T` as a struct that adopts the `Accounts` trait (e.g.
`Context<AddMovieReviewAccounts>`). Through this `context` argument the

> instruction can then access:

Reviewer: I think we can delete most of this section. It duplicates the descriptions of each Context item we have in the next section.

Replace it with:

> The first argument of every instruction handler must be `Context`. Context takes a generic of your Accounts struct, eg,  if `AddMovieReview` was the struct holding the accounts, the context for the add_movie_review() function would be `Context<AddMovieReview>`,

Then add a ContentBox (see CONTRIBUTING) with:

> Yes, the Accounts struct is typically named the same thing as the instruction handler, just in TitleCase. Eg, the struct with the accounts for add_movie_review() is called AddMovieReview! 

BTW I should have asked you to do the same in `intro-to-anchor` during #349, since that chapter does the same thing - pastes the Rust code describing the struct, then describes the struct straight afterwards!|

Author: Couldn't find ContentBox in the CONTRIBUTING. Do you mean callout by any chance?

Reviewer: Callout! Yes.

Author: done

<callout type="info" title="Naming convention">
  Yes, the Accounts struct is typically named the same thing as the instruction handler, just in TitleCase. Eg, the struct with the accounts for add_movie_review() is called AddMovieReview!
</callout>

Through this context argument the instruction can then access:

- The accounts passed into the instruction (`ctx.accounts`)
- The program ID (`ctx.program_id`) of the executing program
- The remaining accounts (`ctx.remaining_accounts`). The `remaining_accounts` is
  a vector that contains all accounts that were passed into the instruction but
  are not declared in the `Accounts` struct.
- The bumps for any PDA accounts in the `Accounts` struct (`ctx.bumps`)
- The seeds for any PDA accounts in tha `Accounts` struct (`ctx.seeds`)

> The design of Contexts can be different across different programs to serve
> their purpose; and the name of the context could be anything (not limited to
> Context) to better reflect it's usage. This example is to help understand how
> contexts work in Anchor.

### Define instruction accounts

The `Accounts` trait:

- Defines a structure of validated accounts for an instruction

> - Makes accounts accessible through an instruction's `Context`

Reviewer: Can you please go through and fix usage of 'instruction handler' (for the function) versus 'instruction' (for the input to the function)? See CONTRIBUTING.md

Author: yes sure

Author: did it

`Context<AddMovieReviewAccounts>`). Through this context argument the
instruction can then access:

> Through this context argument the instruction can then access:

Reviewer: ```suggestion
Through this `context `argument the instruction can then access:
```

- The accounts passed into the instruction (`ctx.accounts`)
- The program ID (`ctx.program_id`) of the executing program
- The remaining accounts (`ctx.remaining_accounts`). The `remaining_accounts` is
  a vector that contains all accounts that were passed into the instruction but
  are not declared in the `Accounts` struct.
- The bumps for any PDA accounts in the `Accounts` struct (`ctx.bumps`)
- The seeds for any PDA accounts in tha `Accounts` struct (`ctx.seeds`)

> The design of Contexts can be different across different programs to serve
> their purpose; and the name of the context could be anything (not limited to
> Context) to better reflect it's usage. This example is to help understand how
> contexts work in Anchor.

### Define instruction accounts

The `Accounts` trait defines a structure of validated accounts. Structs that
adopt this trait specify the accounts required for an instruction. These
accounts are then accessible through an instruction's `Context`, eliminating the

> need for manual account handling and deserialization.

Reviewer: The old version of this was needlessly complex. Let's make it way better. 

> One of the benefits of Solana compared to older blockchains is that transactions that involve different accounts don't stop each other from being processed. Alice can send USDC to Bob while Charles makes a transaction to Diane's lending protocol, and neither transaction will block the other. This means that the incoming instructions must include a list of the accounts that will be used in the instruction. 

> The `Accounts` trait is applied to a `struct` that specifies the accounts required for each instruction, as well as any additional requirements for those accounts - for example requiring a particular account to have signed the transaction. These accounts are then accessible through an instruction's `context` parameter.

Just a suggestion. Feel free to write your own!

Author: sure, I can try

}

#[derive(Accounts)]
pub struct InstructionAccounts<'info> {
    #[account(
        init,
        payer = user,

>         space = (DISCRIMINATOR as usize) + AccountStruct::INIT_SPACE

Reviewer: Just make DISCRIMINATOR a usize in the first place! Also call it ANCHOR_DISCRIMINATOR since that's what we've used elsewhere.

  it (sets its account discriminator)
- `payer` - specifies the payer for the account initialization to be the `user`
  account defined in the struct

> - `space`- specifies that the space allocated for the account.
>   - `DISCRIMINATOR` is a constant defined in the `AccountStruct` implementation.
>     The first 8 bytes of an account are reserved for the discriminator.
>     Discriminators are used to differentiate between different account types It
>     is of type `u8`.
>   - `INIT_SPACE` is a constant defined in the `AccountStruct` implementation. It
>     is a more convenient and readable way to specify the space required for the
>     account. It includes the total space (in bytes) required for the account. It
>     is of type `usize`.
>   - The very need of using this `space` constraint can be eliminated by using
>     `#[derive(InitSpace)]` macro. We'll see how to use that further in this
>     lesson.

Reviewer: ```suggestion
- `space`- the space allocated on the blockchain to store the account.
  - `DISCRIMINATOR` is the first 8 bytes of an account, which Anchor uses to save the type of the account.
   - `AccountStruct::INIT_SPACE` is the total size of space required for all the items in the
    `AccountStruct`. 
```

```rust
#[derive(Accounts)]
pub struct InstructionAccounts {
    #[account(init,
        payer = user,
        space = (DISCRIMINATOR as usize) + AnchorStruct::INIT_SPACE
    )]
    pub account_name: Account<'info, AccountStruct>,
    ...
}

#[account]
#[derive(InitSpace)]
pub struct AccountStruct {
    data: u64
}

> const DISCRIMINATOR: u8 = 8;

Reviewer: `usize` as mentioned!

```

The `#[account]` attribute ensures that it can be used as an account in
`InstructionAccounts`.

When the `account_name` account is initialized:

- The first 8 bytes is set as the `AccountStruct` discriminator using the
  `DISCRIMINATOR` constant.
- The data field of the account will match `AccountStruct`
- The account owner is set as the `programId` from `declare_id`
- The account account struct is made with the `[derive(InitSpace)]` macro which
  provides `INIT_SPACE` constant to the `AccountStruct`. It calculates the space
  required by the account based on the account's data-structure types.

> It is considered a good practice to use the `#[derive(InitSpace)]` macro which

> > makes the code more readable and maintainable.

Reviewer: We can delete this, we've already covered the topic.

pub struct AccountStruct {
    data: u64
}

const DISCRIMINATOR: usize = 0;
const DATA_SIZE: usize = 8;

impl Space for AccountStruct {
    const INIT_SPACE: usize = DISCRIMINATOR + DATA_SIZE;

> }

Reviewer: Remove DATA_SIZE, use https://www.anchor-lang.com/docs/space#the-init-space-macro then space = DISCRIMINATOR + INIT_SPACE

// Validate incoming accounts for instructions
#[derive(Accounts)]
pub struct InstructionAccounts<'info> {
    #[account(init,
        payer = user,
        space = (DISCRIMINATOR as usize) + AccountStruct::INIT_SPACE
    )]
    pub account_name: Account<'info, AccountStruct>,
    #[account(mut)]
    pub user: Signer<'info>,
    pub system_program: Program<'info, System>,
}

// Define custom program account type
#[account]
#[derive(InitSpace)]
pub struct AccountStruct {
    data: u64
}

const DISCRIMINATOR: u64 = 8;
```

#### Key takeaways:

- The whole program structure can be broadly divided into three parts:
  1. Account constraints: define the accounts required for the instructions, as
     well as rules to apply to them - e.g., whether they need to sign the
     transaction, if they should be created on demand, how addresses for PDAs,
     etc.
  2. Instruction handlers: implement program logic, as functions inside
     the`#[program]` module.
  3. Accounts: define the format used for data accounts.

> 

Reviewer: great addition! 👍

pub struct AccountStruct {
    data: u64
}

// Space calculation for account the AccountStruct
const DISCRIMINATOR: usize = 0;

> const DATA_SIZE: usize = 8;

Reviewer: See above.

impl Space for  AccountStruct {
    const INIT_SPACE: usize = DISCRIMINATOR + DATA_SIZE;
}
```

#### Key takeaways:

> - The whole program structure can be broadly divided into three parts:
>   1. Instruction Logic: where the instruction handlers are defined (i.e., the
>      `#[program]` module).
>   2. Instruction Contexts: where the accounts required for the instructions are
>      configured.
>   3. Account Types: where the data structure of the accounts are defined.
> 

Reviewer: ```suggestion
- The whole program structure can be broadly divided into three parts:
  1. Account constraints: define the accounts required for the instructions, as well as rules to apply to them - e.g., whether they need to sign the transaction, if they should be created on demand, how addresses for PDAs, etc.  
  2. Instruction handlers: implement program logic, as functions inside the`#[program]` module.
  3. Accounts: define the format used for data accounts. 
```

Anchor uses the term 'constraints' so we do too https://www.anchor-lang.com/docs/the-accounts-struct#constraints

Reviewer: Also reordered this - constraints, then function, then account structs since that's the order they get executed.

Author: sure, I'll incorporate all this in the docs
